File: Beam/Source/QueriesTests/InterruptableQueryTester.cpp
#include <doctest/doctest.h>
#include "Beam/Queries/InterruptableQuery.hpp"
#include "Beam/SerializationTests/ValueShuttleTests.hpp"
#include "Beam/Utilities/ToString.hpp"

using namespace Beam;
using namespace Beam::Tests;

TEST_SUITE("InterruptableQuery") {
  TEST_CASE("default_constructor") {
    auto query = InterruptableQuery();
    REQUIRE(query.get_interruption_policy() == InterruptionPolicy::BREAK_QUERY);
  }

  TEST_CASE("interruption_policy_constructor") {
    auto break_query = InterruptableQuery(InterruptionPolicy::BREAK_QUERY);
    REQUIRE(
      break_query.get_interruption_policy() == InterruptionPolicy::BREAK_QUERY);
    auto recover_query = InterruptableQuery(InterruptionPolicy::RECOVER_DATA);
    REQUIRE(recover_query.get_interruption_policy() ==
      InterruptionPolicy::RECOVER_DATA);
    auto continue_query =
      InterruptableQuery(InterruptionPolicy::IGNORE_CONTINUE);
    REQUIRE(continue_query.get_interruption_policy() ==
      InterruptionPolicy::IGNORE_CONTINUE);
  }

  TEST_CASE("set_interruption_policy") {
    auto query = InterruptableQuery();
    REQUIRE(
      query.get_interruption_policy() != InterruptionPolicy::IGNORE_CONTINUE);
    query.set_interruption_policy(InterruptionPolicy::IGNORE_CONTINUE);
    REQUIRE(
      query.get_interruption_policy() == InterruptionPolicy::IGNORE_CONTINUE);
    query.set_interruption_policy(InterruptionPolicy::RECOVER_DATA);
    REQUIRE(
      query.get_interruption_policy() == InterruptionPolicy::RECOVER_DATA);
    query.set_interruption_policy(InterruptionPolicy::BREAK_QUERY);
    REQUIRE(query.get_interruption_policy() == InterruptionPolicy::BREAK_QUERY);
  }

  TEST_CASE("stream") {
    auto query = InterruptableQuery(InterruptionPolicy::BREAK_QUERY);
    REQUIRE(to_string(query) == "BREAK_QUERY");
    test_round_trip_shuttle(query);
  }
}

File: WebApi/tests/queries/interruptable_query_tester.ts
import { describe, it } from 'node:test';
import * as assert from 'node:assert';
import { InterruptableQuery } from
  '../../source/queries/interruptable_query';
import { InterruptionPolicy } from
  '../../source/queries/interruption_policy';

describe('InterruptableQuery', () => {
  it('default_constructor', () => {
    assert.strictEqual(new InterruptableQuery().interruptionPolicy,
      InterruptionPolicy.BREAK_QUERY);
  });

  it('interruption_policy', () => {
    const query = new InterruptableQuery();
    query.interruptionPolicy = InterruptionPolicy.RECOVER_DATA;
    assert.strictEqual(
      query.interruptionPolicy, InterruptionPolicy.RECOVER_DATA);
  });

  it('equals', () => {
    const query = new InterruptableQuery(InterruptionPolicy.IGNORE_CONTINUE);
    assert.ok(query.equals(
      new InterruptableQuery(InterruptionPolicy.IGNORE_CONTINUE)));
    assert.ok(!query.equals(new InterruptableQuery()));
  });

  it('to_string', () => {
    assert.strictEqual(new InterruptableQuery().toString(), 'BREAK_QUERY');
  });

  it('round_trip', () => {
    const query = new InterruptableQuery(InterruptionPolicy.RECOVER_DATA);
    const json = query.toJson();
    assert.strictEqual(
      json.interruption_policy, InterruptionPolicy.RECOVER_DATA);
    assert.ok(InterruptableQuery.fromJson(json).equals(query));
  });
});
